bulk_indexer.go

```diff
@@ -240,9 +272,128 @@ func (b *bulkIndexer) Flush(ctx context.Context) (BulkIndexerResponseStat, error
 	if err := jsoniter.NewDecoder(res.Body).Decode(&resp); err != nil {
 		return resp, fmt.Errorf("error decoding bulk response: %w", err)
 	}
+
+	// Only run the retry logic if document retries are enabled
+	if b.maxDocumentRetry > 0 {
+		buf := make([]byte, 1024)
+
+		// Eliminate previous retry counts that aren't present in the bulk
+		// request response.
+		for k := range b.retryCounts {
+			found := false
+			for _, res := range resp.FailedDocs {
+				if res.Position == k {
+					found = true
+					break
+				}
+			}
+			if !found {
+				// Retried request succeeded, remove from retry counts
+				delete(b.retryCounts, k)
+			}
+		}
+
+		tmp := resp.FailedDocs[:0]
+		for _, res := range resp.FailedDocs {
+			if res.Status == http.StatusTooManyRequests {
+				// there are two lines for each document:
+				// - action
+				// - document
+				//
+				// Find the document by looking up the newline separators.
+				// First the newline (if exists) before the 'action' then the
+				// newline at the end of the 'document' line.
+				startln := res.Position * 2
+				endln := startln + 2
+
+				// Increment 429 count for the positions found.
+				count := b.retryCounts[res.Position] + 1
+				// check if we are above the maxDocumentRetry setting
+				if count > b.maxDocumentRetry {
+					// do not retry, return the document as failed
+					tmp = append(tmp, res)
+					continue
+				}
+
+				// Since some items may have succeeded, counter positions need
+				// to be updated to match the next current buffer position.
+				b.retryCounts[b.itemsAdded] = count
+
+				if b.gzipw != nil {
+					gr, err := gzip.NewReader(bytes.NewReader(b.copyBuf))
```

Comment:
It looks like we're decompressing and scanning through the request body for every failed item. Should we be decompressing once, and scanning through incrementally? e.g. if you have failures at index 1 and 8, then first skip to the index 1, then skip from index 1 to index 8 without decompressing and reprocessing up to index 1.

Reply:
I don't mean decompress the whole payload at once, I meant create the decompressor once and stream through the results. AFAICS there's no need to call `gzip.NewReader` more than once.

Instead you could lazily create it when seeing the first 429 result, wrap it with [bufio.NewScanner](https://pkg.go.dev/bufio#NewScanner), and stream through the lines, skipping non-429 result lines as needed.